pennylane/ops/op_math/controlled_ops.py

```diff
@@ -334,3 +340,894 @@ def compute_matrix():  # pylint: disable=arguments-differ
 
     def _controlled(self, wire):
         return CCZ(wires=wire + self.wires)
+
+
+class CRX(ControlledOp):
+    r"""The controlled-RX operator
+
+    .. math::
+
+        \begin{align}
+            CR_x(\phi) &=
+            \begin{bmatrix}
+            & 1 & 0 & 0 & 0 \\
+            & 0 & 1 & 0 & 0\\
+            & 0 & 0 & \cos(\phi/2) & -i\sin(\phi/2)\\
+            & 0 & 0 & -i\sin(\phi/2) & \cos(\phi/2)
+            \end{bmatrix}.
+        \end{align}
+
+    **Details:**
+
+    * Number of wires: 2
+    * Number of parameters: 1
+    * Number of dimensions per parameter: (0,)
+    * Gradient recipe: The controlled-RX operator satisfies a four-term parameter-shift rule
+      (see Appendix F, https://doi.org/10.1088/1367-2630/ac2cb3):
+
+      .. math::
+
+          \frac{d}{d\phi}f(CR_x(\phi)) = c_+ \left[f(CR_x(\phi+a)) - f(CR_x(\phi-a))\right] - c_- \left[f(CR_x(\phi+b)) - f(CR_x(\phi-b))\right]
+
+      where :math:`f` is an expectation value depending on :math:`CR_x(\phi)`, and
+
+      - :math:`a = \pi/2`
+      - :math:`b = 3\pi/2`
+      - :math:`c_{\pm} = (\sqrt{2} \pm 1)/{4\sqrt{2}}`
+
+    Args:
+        phi (float): rotation angle :math:`\phi`
+        wires (Sequence[int]): the wire the operation acts on
+        id (str or None): String representing the operation (optional)
+    """
+    num_wires = 2
+    """int: Number of wires that the operation acts on."""
+
+    num_params = 1
+    """int: Number of trainable parameters that the operator depends on."""
+
+    ndim_params = (0,)
+    """tuple[int]: Number of dimensions per trainable parameter that the operator depends on."""
+
+    name = "CRX"
+    basis = "X"
+    grad_method = "A"
+    parameter_frequencies = [(0.5, 1.0)]
+
+    def __init__(self, phi, wires, id=None):
+        super().__init__(qml.RX(phi, wires=wires[1:]), control_wires=wires[0], id=id)
+
+    @classmethod
+    def _unflatten(cls, data, metadata):
+        base = data[0]
+        control_wires = metadata[0]
+        return cls(*base.data, wires=control_wires + base.wires)
+
+    def label(self, decimals=None, base_label=None, cache=None):
+        return super().label(decimals=decimals, base_label=base_label or "RX", cache=cache)
+
+    @staticmethod
+    def compute_matrix(theta):  # pylint: disable=arguments-differ
```

Comment:
So we overrode the `compute_matrix` method in `CZ` and `CY`, since just creating the required array was so much quicker than all the logic in `Controlled.matrix`.  So I think the question here is whether or not this `compute_matrix` implementation is worth overriding `Controlled.matrix`.


We can technically get all the required behaviour by simply inheriting.  Then we have less code to maintain.  Is it worth having a specialized implementation here?

Reply:
What's the performance overhead on using the default behaviour? Is that a concern? If not, I can probably remove those because they literally produce the same result.

Reply:
So when I overrode `CRX.matrix` to make sure it's still using `CRX.compute_matrix`:

```
op1 = qml.CRX(1.2, wires=(0,1))
op2 = qml.ops.op_math.Controlled(qml.RX(1.2,0), 1)
```

![Screenshot 2024-01-18 at 9 36 09 AM](https://github.com/PennyLaneAI/pennylane/assets/6364575/f91375e0-9a54-41da-9359-ed60ae96a26c)


For comparison, when I override `CY.matrix` so it actually uses `CY.compute_matrix`, the difference is

![Screenshot 2024-01-18 at 9 39 56 AM](https://github.com/PennyLaneAI/pennylane/assets/6364575/d7f792e8-15ae-4a8e-b98f-cf4533131ad2)

So the general path with more logic is indeed a little bit slower.

For the non-matrix ops, this speed-up might just be worth it, since it's a factor of 1,000. 

For the rotation gates, I'm not sure the performance difference is substantial enough to warrant the duplicated code.

Reply:
As discussed, deferring this to a separate PR at some point in the future.